processing/src/main/java/org/apache/druid/query/operator/WindowOperatorQueryQueryToolChest.java

```diff
@@ -158,42 +177,77 @@ public Sequence run(
             queryPlus.withQuery(query.withOverriddenContext(ImmutableMap.of("unravel", false))),
             responseContext
         );
+        final ResultSerializationMode serializationMode = query.context().getEnum(
+            ResultSerializationMode.CTX_SERIALIZATION_PARAMETER,
+            ResultSerializationMode.class,
+            ResultSerializationMode.ROWS
+        );
+        switch (serializationMode) {
+          case ROWS:
+            return asRows(baseSequence, query);
+          case FRAMES:
+            return asFrames(baseSequence);
+          default:
+            throw DruidException.defensive("Serialization mode[%s] not supported", serializationMode);
+        }
+      }
 
-        final RowSignature rowSignature = query.getRowSignature();
-        return baseSequence.flatMap(
-            rac -> {
-              List<Object[]> results = new ArrayList<>(rac.numRows());
-
-              ColumnAccessor[] accessors = new ColumnAccessor[rowSignature.size()];
-              int index = 0;
-              for (String columnName : rowSignature.getColumnNames()) {
-                final Column column = rac.findColumn(columnName);
-                if (column == null) {
-                  final ColumnType columnType = rowSignature
-                      .getColumnType(columnName)
-                      .orElse(ColumnType.UNKNOWN_COMPLEX);
-
-                  accessors[index] = new NullColumn.Accessor(columnType, rac.numRows());
-                } else {
-                  accessors[index] = column.toAccessor();
-                }
-                ++index;
-              }
+      return baseQueryRunner.run(queryPlus, responseContext);
+    }
 
-              for (int i = 0; i < rac.numRows(); ++i) {
-                Object[] objArr = new Object[accessors.length];
-                for (int j = 0; j < accessors.length; j++) {
-                  objArr[j] = accessors[j].getObject(i);
-                }
-                results.add(objArr);
+    /**
+     * Translates Sequence of RACs to a Sequence of Object[]
+     */
+    private static Sequence asRows(final Sequence<RowsAndColumns> baseSequence, final WindowOperatorQuery query)
+    {
+      final RowSignature rowSignature = query.getRowSignature();
+      return baseSequence.flatMap(
+          rac -> {
+            List<Object[]> results = new ArrayList<>(rac.numRows());
+
+            ColumnAccessor[] accessors = new ColumnAccessor[rowSignature.size()];
+            int index = 0;
+            for (String columnName : rowSignature.getColumnNames()) {
+              final Column column = rac.findColumn(columnName);
+              if (column == null) {
```

Comment:
Druid has a lot of places where the absence of a column is treated as all nulls, so leaving it as is is a less risky proposition. Can this happen if one of the segments doesn't contain a row, while the others do? 

That said, this portion has been refactored as-is from the pre-existing code. I'll leave it to you to decide if this is applicable or not.

Reply:
around here it seems like we already had that in the signature; but if the rac doesn't know about that column I think its already inconsistent for some reason; it might be due to some other issue...

> That said, this portion has been refactored as-is from the pre-existing code. I'll leave it to you to decide if this is applicable or not.

if we could add an exception and all tests pass I think we are better of with that  :crossed_fingers: